homeassistant/components/flickelectric/sensor.py

```diff
@@ -0,0 +1,174 @@
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_CLIENT_ID): cv.string,
+        vol.Optional(CONF_CLIENT_SECRET): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Flick sensor."""
+    websession = async_get_clientsession(hass)
+
+    async_add_entities(
+        [
+            FlickPricingSensor(
+                hass.loop,
+                websession,
+                config[CONF_USERNAME],
+                config[CONF_PASSWORD],
+                config.get(CONF_CLIENT_ID),
+                config.get(CONF_CLIENT_SECRET),
+            )
+        ],
+        True,
+    )
+
+
+class FlickPricingSensor(Entity):
+    """Implementation of the Flick Pricing sensor."""
+
+    def __init__(
+        self,
+        loop,
+        websession: aiohttp.ClientSession,
+        username: str,
+        password: str,
+        client_id: str,
+        client_secret: str,
+    ):
+        """Initialize the sensor."""
+
+        self.loop = loop
+        self._username: str = username
+        self._password: str = password
+        self.websession: aiohttp.ClientSession = websession
+
+        if client_id:
+            self._client_id: str = client_id
+        else:
+            self._client_id: str = "le37iwi3qctbduh39fvnpevt1m2uuvz"
+
+        if client_secret:
+            self._client_secret: str = client_secret
+        else:
+            self._client_secret: str = "ignwy9ztnst3azswww66y9vd9zt6qnt"
+
+        self._name: str = FRIENDLY_NAME
+        self._state = None
+        self._unit_of_measurement: str = UNIT_NAME
+        self._attributes = {
+            ATTR_ATTRIBUTION: ATTRIBUTION,
+            ATTR_FRIENDLY_NAME: FRIENDLY_NAME,
+        }
+
+        self._token = None
+        self._token_next_refresh = dt.now()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    async def _refresh_token(self):
+        try:
+            with async_timeout.timeout(60):
+
+                data = aiohttp.FormData()
+                data.add_field("grant_type", "password")
+                data.add_field("client_id", self._client_id)
+                data.add_field("client_secret", self._client_secret)
+                data.add_field("username", self._username)
+                data.add_field("password", self._password)
+
+                response = await self.websession.post(_AUTH_URL, data=data)
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Reply:
Done, extracted into [PyFlick](https://pypi.org/project/PyFlick/0.0.2/) ([Github](https://github.com/ZephireNZ/PyFlick))